### PR TITLE
api, vmware: Allow VM setting/detail for disk controller (root/data) to override template details

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR allows VM details wrt root & data disk controller passed via details parameter of deployVMCmd to take precedence over template settings, i.e., the order of precedence for determining VM disk controller is : VM details passed during deployment > template settings > global setting `vmware.root.disk.controller` .

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@Pearl1594 is this ready for review?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -4442,6 +4421,42 @@ public UserVmVO doInTransaction(TransactionStatus status) throws InsufficientCap
...
+    private void updateVMDiskController(UserVmVO vm, Map<String, String> customParameters, GuestOSVO guestOS) {
+        // If hypervisor is vSphere and OS is OS X, set special settings.
+        if (guestOS.getDisplayName().toLowerCase().contains("apple mac os")) {
+            vm.setDetail(VmDetailConstants.SMC_PRESENT, "TRUE");
+            vm.setDetail(VmDetailConstants.ROOT_DISK_CONTROLLER, "scsi");
+            vm.setDetail(VmDetailConstants.DATA_DISK_CONTROLLER, "scsi");
+            vm.setDetail(VmDetailConstants.FIRMWARE, "efi");
+            s_logger.info("guestOS is OSX : overwrite root disk controller to scsi, use smc and efi");
+        } else {
+            String rootDiskControllerSetting = customParameters.get(VmDetailConstants.ROOT_DISK_CONTROLLER);
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4433]: code lgtm

**Review by weizhouapache — APPROVED** [review]

**Review by harikrishna-patnala — APPROVED** [review]
LGTM

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -4442,6 +4421,42 @@ public UserVmVO doInTransaction(TransactionStatus status) throws InsufficientCap
...
+                vm.setDetail(VmDetailConstants.DATA_DISK_CONTROLLER, dataDiskControllerSetting);
+            }
+
+            String controllerSetting = StringUtils.defaultIfEmpty(_configDao.getValue(Config.VmwareRootDiskControllerType.key()),
+                    Config.VmwareRootDiskControllerType.getDefaultValue());
+
+            // Don't override if VM already has root/data disk controller detail
+            if (vm.getDetail(VmDetailConstants.ROOT_DISK_CONTROLLER) == null) {
+                vm.setDetail(VmDetailConstants.ROOT_DISK_CONTROLLER, controllerSetting);
+            }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4449]: In UserVmVO we could create a method like `setDetailIfNull`. This way we can centralize the validation and reduce a bit of code.

**Comment by rohityadavcloud** [comment]
Is this essential for 4.17 @nvazquez @Pearl1594 ?

**Comment by Pearl1594** [comment]
@rohityadavcloud this isn't essential for 4.17.0 as this has always been the behavior and there also exists a workaround, where users can change the template setting for root disk controller. So IMO this can go into 4.17.1

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

**Review by sureshanaparti — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@Pearl1594 will you apply @GutoVeronezi's suggestion?

**Comment by harikrishna-patnala** [comment]
@Pearl1594 This also needs base branch change to 4.17, I see this is targetted to 4.17.1

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@Pearl1594 can you address Guto's review comment. Other than that PR looks all good.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 2 above)
Comment (Pearl1594) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:4449]: I am not sure if it's worth doing this at this point, as it's not merely checking if it's null and setting it to some value.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmaware-70u3

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-70u3

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-70u3

**Comment by DaanHoogland** [comment]
> Test 	Result 	Time (s) 	Test File
> test_deploy_vm_from_iso 	`Error` 	7.84 	test_deploy_vm_iso.py

@Pearl1594 , this remaining error seems to be consistent. Can you look at it?

**Review by shwstppr — COMMENTED** [review]
@Pearl1594 failing vm from iso test is due to `validateRootDiskResize` should we call it only for templates and not iso?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-70u3

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Review by harikrishna-patnala — APPROVED** [review]
Tested the VM root disk controller settings at different places. 
1. With only global setting value
2. With global setting and template setting
3. With global, template and VM deployment
4. Stopped the VM, changed VM setting in the above cases and started the VM

I've checked the VM settings on vCenter, in all the above cases I observed the precedence in the order of 
VM setting > Template setting > Global setting

**Pull request merged**